### ISSUE TITLE
create_task() recommendation seems like bad practice

### ISSUE THREAD

**Issue by Dreamsorcerer** [python/cpython#104091]
# Documentation

My understanding of asyncio is that all tasks etc. should be `await`ed on. If this is not done, then exceptions in tasks will never be seen and asyncio will give an "exception was not retrieved" warning.

The documentation currently recommends this code for a fire-and-forget task creation:
```py
background_tasks = set()
for i in range(10):
    task = asyncio.create_task(some_coro(param=i))
    background_tasks.add(task)
    task.add_done_callback(background_tasks.discard)
```
https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task

This seems like bad practice as the tasks are never awaited (another problem might be if this is copied inside a function, won't the set get garbage collected after the function ends?). To my knowledge, there is no trivial way to do fire-and-forget tasks using stdlib and I have always recommended users to try the aiojobs library to handle this. e.g. In the aiohttp documentation:
https://docs.aiohttp.org/en/latest/web_advanced.html#web-handler-cancellation
(Scroll down to the second warning, and the paragraph preceding it).

I feel that this example should probably be scrapped and possibly a recommendation to use aiojobs for these tasks may be better.

I'd also suggest including the text of both error messages ("exception not retrieved" and "Task was destroyed"), so that people putting that error into a search engine are likely to end up on this page.

**Comment by carljm** [comment]
I think there's a misunderstanding here about what `create_task` does. Which is understandable: the name `create_task` is arguably misleading by omission. It isn't just a pure function that creates and returns a `Task` instance; it creates a task _and schedules it to the event loop_. (See [the documentation](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task).)

This means the coroutine doesn't need to be explicitly awaited, because the event loop will take care of ensuring that it gets waited on. The event loop also keeps a reference to the Task object, which keeps a reference to the coroutine, so nothing will be garbage collected until it's done running.

So this documented recommendation for fire-and-forget task creation does work fine, and is not bad practice (except inasmuch as [TaskGroups](https://docs.python.org/3/library/asyncio-task.html#task-groups) are probably a better option in most cases.)

**Comment by Dreamsorcerer** [comment]
> This means the coroutine doesn't need to be explicitly awaited, because the event loop will take care of ensuring that it gets waited on.

No, it ensures it executes, not that it get `await`ed on. You can't get an "exception not retrieved" if the code hasn't run to create an exception...

Simple reproducer:
```
>>> import asyncio
>>> async def foo():
...  raise RuntimeError()
... 
>>> async def main():
...  asyncio.create_task(foo())
...  await asyncio.sleep(1)
... 
>>> asyncio.run(main())
Task exception was never retrieved
...
```

Well written code should never see "Task exception was never retrieved".

> So this documented recommendation for fire-and-forget task creation does work fine, and is not bad practice (except inasmuch as [TaskGroups](https://docs.python.org/3/library/asyncio-task.html#task-groups) are probably a better option in most cases.)

TaskGroups can only be used for this if you create some kind of global TaskGroup for your application that you keep around till the application ends. This can work, but in something like a server application, I think you'd only get the exceptions when the server shuts down, which could be weeks after the exception happened. So, for server applications, it'd still be best to use aiojobs or similar.

**Comment by Dreamsorcerer** [comment]
You'll see in aiojobs, that it awaits on the task even when it has nothing to do, explicitly to avoid that warning:
https://github.com/aio-libs/aiojobs/blob/master/aiojobs/_scheduler.py#L161

That code was written by asvetlov, one of the asyncio code owners, so I'm pretty sure it's the right thing to do.

**Comment by carljm** [comment]
Sorry, I did misunderstand your report.

The name "fire-and-forget" specifically suggests that you don't care if it succeeds or fails. If you are going to check the result/exception, then it isn't really fire and forget.

I'm not sure if we should be recommending a specific third-party library here, but it may make sense to add a note in the documentation that fire-and-forget means you won't know if it succeeded or failed, and you may get unretrieved exception warnings in case of failure.

I'll reopen this so asyncio experts can weigh in on what, if anything, should be improved here.

**Comment by gvanrossum** [comment]
The coroutine running in the background task could contain a try/except clause that catches all exceptions and logs them. You could easily make a decorator that does this for your application so you don't have to repeat the try/except/log boilerplate in every background task you write.

We're working on additional flexibility for TaskGroup that will make this pattern simpler. (See e.g. https://discuss.python.org/t/revisiting-persistenttaskgroup-with-kotlins-supervisorscope/18417)

**Comment by Dreamsorcerer** [comment]
> We're working on additional flexibility for TaskGroup that will make this pattern simpler. (See e.g. https://discuss.python.org/t/revisiting-persistenttaskgroup-with-kotlins-supervisorscope/18417)

Right, I think if TaskGroup starts supporting logging exceptions as they occur, then it should be able to replace aiojobs.Scheduler.

So, I think we should be recommending a long-lived TaskGroup for fire-and-forget tasks then, instead of the currently documented workaround.

It might also be worth having a `TaskGroup.close()` (and `.start()`?) to support this use case too. e.g. In startup/shutdown hooks:
```
async def on_startup(app):
    app["tg"] = asyncio.TaskGroup()
async def on_shutdown(app):
    await app["tg"].close()
```

In aiohttp, we have cleanup_ctx, where we'd use `async with`, but I'm sure there are frameworks without that kind of hook.

**Comment by gvanrossum** [comment]
TaskGroup itself won't be logging exceptions, you will have to write your own exception handler or done-callback for that.

For TaskGroup.start() and .close() you can call `__aenter__()` and `__aexit__()` directly (these are public APIs).

I don't see an action item here.

**Comment by gvanrossum** [comment]
> But, it will default to `asyncio.get_running_loop().call_exception_handler()`, right? That's the same as aiojobs, so it should still work as a simple replacement.

I would like to push back on this a little. Why do you want it to call `call_exception_handler()`? Looking over the asyncio source code, that function is generally only called when there is no alternative, e.g. when an exception is caught in a finalizer.

The regTaskGroup doesn't call it when a task crashes

**Comment by gvanrossum** [comment]
Whoops, closed by mistake. Will resume in a bit.

**Comment by gvanrossum** [comment]
> But, it will default to `asyncio.get_running_loop().call_exception_handler()`, right? That's the same as aiojobs, so it should still work as a simple replacement.

I would like to push back on this a little. Why do you want it to call `call_exception_handler()`? Looking over the asyncio source code, that function is generally only called when there is no alternative, e.g. when an exception is caught in a finalizer.

The regular `TaskGroup` doesn't call it when a task fails, so why should the variant for background tasks call it? I still think that there are already plenty of other ways for the user to arrange for task failures to be logged, e.g. by wrapping the task's coroutine in a handler that logs the exception, or by adding a done-callback to do the same. (@achimnol seems to agree, his latest `Supervisor` design does not do this.)

(Though TBH I'm not 100% sold on my own argument here, so maybe you have a good argument?)

> Fair enough, I think generally people discourage use of calling magic methods directly, so I was thinking users will be more comfortable with a public (I mean, no `_` prefix) method. I think that'd also be consistent with other objects too (e.g. file-like objects have `.close()` in addition to `__exit__()`). Either way, it'll work though.

Technically, a dunder is not "private" or "protected" in the sense that a single or double underscore conveys. For files, `close()` long predated `__exit__()`. I'm not too keen on violating TOOWTDI here. But maybe the `Supervisor` class shouldn't be an async context manager, solving that problem.

Regarding the original doc example, I don't necessarily like making all examples industrial-strength, because (to me) the purpose of an example is understanding, not a copy-pastable solution. (You end up with very verbose examples if you strive for the latter.) I'd be okay with a small update to that particular example though that reminds the reader that if the task fails, no exception will be printed anywhere. The user can then decide what to do about that.

**Comment by Dreamsorcerer** [comment]
> I would like to push back on this a little. Why do you want it to call `call_exception_handler()`?

It was mentioned at https://discuss.python.org/t/revisiting-persistenttaskgroup-with-kotlins-supervisorscope/18417/12
So, I thought it was already part of the plan.

I was thinking that aiohttp installs it's own default handler, but now that I look I'm not so sure. I'll give a try later, but basically I'd expect to atleast have exceptions appear on stdout if no log handler is provided.

> The regular `TaskGroup` doesn't call it when a task fails, so why should the variant for background tasks call it?

In that case the exception is reraised and can then be handled by other code, right? Whereas a background task could only be reraised when the application ends. So, in this scenario, I'd expect the exception to be handled immediately and not raise anything on shutdown.

> I still think that there are already plenty of other ways for the user to arrange for task failures to be logged, e.g. by wrapping the task's coroutine in a handler that logs the exception, or by adding a done-callback to do the same. (@achimnol seems to agree, his latest `Supervisor` design does not do this.)
> 
> (Though TBH I'm not 100% sold on my own argument here, so maybe you have a good argument?)

My argument is really just to do the expected thing by default. i.e. A user who creates this Supervisor without much knowledge to best practices should not then end up spending hours or days trying to figure out why their code is not behaving correctly, simply because they never even knew an exception was happening in their code.

It would also just be convenient, not having to add (and remember to add) a custom decorator or whatever everytime you quickly create a task to throw into the Supervisor.

> But maybe the `Supervisor` class shouldn't be an async context manager, solving that problem.

I'm not too bothered either way, as I say, we have cleanup_ctx in aiohttp, so it'll either look like:
```
async Supervisor() as s:
    app["supervisor"] = s
    yield
```
or
```
app["supervisor"] = Supervisor()
yield
await app["supervisor"].close()
```

**Comment by gvanrossum** [comment]
> > I would like to push back on this a little. Why do you want it to call `call_exception_handler()`?
> 
> It was mentioned at https://discuss.python.org/t/revisiting-persistenttaskgroup-with-kotlins-supervisorscope/18417/12 So, I thought it was already part of the plan.

I have to apologize, my Discourse mailing list mode was off for a while and I missed that post (and many others).

But, in https://discuss.python.org/t/asyncio-tasks-and-exception-handling-recommended-idioms/23806/11 it seems @achimnol is reversing course; I found no trace of `call_exception_handler()` in the [code linked from there](https://github.com/achimnol/aiotools/pull/53/files#diff-88c93e83c1838ddba11d3e5e0f15f37f238f75bcb3f910ca3b5160b471ec74f1). (Yes, it's called, but only in a branch that the original author of TaskGroup wasn't even should could be reached, according to the comment.)

> I was thinking that aiohttp installs it's own default handler, but now that I look I'm not so sure. I'll give a try later, but basically I'd expect to atleast have exceptions appear on stdout if no log handler is provided.
> 
> > The regular `TaskGroup` doesn't call it when a task fails, so why should the variant for background tasks call it?
> 
> In that case the exception is reraised and can then be handled by other code, right? Whereas a background task could only be reraised when the application ends. So, in this scenario, I'd expect the exception to be handled immediately and not raise anything on shutdown.
> 
> > I still think that there are already plenty of other ways for the user to arrange for task failures to be logged, e.g. by wrapping the task's coroutine in a handler that logs the exception, or by adding a done-callback to do the same. (@achimnol seems to agree, his latest `Supervisor` design does not do this.)
> > (Though TBH I'm not 100% sold on my own argument here, so maybe you have a good argument?)
> 
> My argument is really just to do the expected thing by default. i.e. A user who creates this Supervisor without much knowledge to best practices should not then end up spending hours or days trying to figure out why their code is not behaving correctly, simply because they never even knew an exception was happening in their code.
> 
> It would also just be convenient, not having to add (and remember to add) a custom decorator or whatever everytime you quickly create a task to throw into the Supervisor.

Okay, you have convinced me. I'd like @achimnol's response.

**Comment by Dreamsorcerer** [comment]
One last point I'd say on the logging front is that for smaller personal projects it's pretty common to not configure logging at all. In that case, it's expected to see exceptions/warnings when you run the application from a terminal or by looking at the web server logs, as they would be sent to stdout/stderr by default.

**Comment by achimnol** [comment]
I've been busy for closing up my company's seris A funding along with several big customer follow-ups. 😞 Recently, I began to look into this issue again, with some rewriting of the aiotools experimentation to make the key concepts in TaskGroup/Supervisor _reusable_ to write various coroutine aggregation patterns.

I agree with that "a user should not end up spending hours or days trying to figure out why their code is not behaving correctly, simply because they never even knew an exception was happening in their code." because I also experienced this many times. 😭

**Comment by willingc** [comment]
Moving this issue status back to To Do as I do not believe that anyone is currently documenting this. But it may be helpful for someone interested in Docs to do so.

**Comment by Jonathan-Landeed** [comment]
+1 for anything that avoids "Future exception was never retrieved".

I was trying to implement a TaskGroup that would have similar behavior to `asyncio.gather(*task_tuple, return_exceptions=True)`. I figured aiotools.PersistentTaskGroup was a good place to start, but the unawaited tasks were very annoying to deal with. The exception handlers in `aiojobs.Scheduler` and `aiotools.Supervisor` also behave very strangely. I'm not sure how they're meant to be used.

Bunch of examples here: https://github.com/python/cpython/issues/101581#issuecomment-2749212395

**Comment by Dreamsorcerer** [comment]
> The exception handlers in `aiojobs.Scheduler` and `aiotools.Supervisor` also behave very strangely. I'm not sure how they're meant to be used.

I'm not too sure what the complication is here? I think the proposed solution is likely to work very similarly. Atleast for aiojobs, which I'm more familiar with, the default behaviour is to call asyncio.AbstractEventLoop.call_exception_handler(), which if not configured to do something else, will just log the exception to stderr. A callable can also be passed in order to use your own custom function instead of using call_exception_handler() if desired.

**Comment by Jonathan-Landeed** [comment]
Maybe I'm just really misinterpreting something, but I can't get the exception_handler to run at all.

This was my attempt at using it to populate an ExceptionGroup.
```
import asyncio
from typing import Any

import aiojobs

MULTIPLE_EXCEPTIONS_MESSAGE = "Multiple exceptions occurred"


async def get_string() -> str:
    return "hello"


async def get_int() -> int:
    return 1


# async def raise_exception() -> typing.Never: # also works
async def raise_exception(s: str, *, fail: bool = True) -> None:
    if fail:
        raise Exception(s)


async def slow_print(s: str) -> None:
    await asyncio.sleep(1)
    print(s)


async def test_scheduler(
    *,
    fail: bool = True,
) -> tuple[str, int, None, None]:
    exceptions = []

    def exception_handler(
        scheduler: "aiojobs.Scheduler",
        context: dict[str, Any],
    ) -> None:
        print("never happens")
        exceptions.append(context["exception"])

    async with aiojobs.Scheduler(exception_handler=exception_handler) as scheduler:
        t1 = await scheduler.spawn(get_string())
        t2 = await scheduler.spawn(get_int())
        t3 = await scheduler.spawn(raise_exception("Raise Exception", fail=fail))
        t4 = await scheduler.spawn(slow_print("Slow Print"))
        await scheduler.spawn(raise_exception("Raise Exception 2", fail=fail))

    if exceptions:
        raise BaseExceptionGroup(
            MULTIPLE_EXCEPTIONS_MESSAGE,
            exceptions,
        )

    await t3.wait()  # Raises exception
    return await t1.wait(), await t2.wait(), None, None
```

**Comment by Dreamsorcerer** [comment]
> Maybe I'm just really misinterpreting something, but I can't get the exception_handler to run at all.

That's because they are caused during the close cycle when exiting the context manager, which seems to result in suppressing the exceptions. Not sure that makes sense, so we should probably change that behaviour. But for now, if you add a sleep before exiting the context, then it does work as expected.

**Comment by Dreamsorcerer** [comment]
> Not sure that makes sense, so we should probably change that behaviour.

Definitely a bug I introduced when added the wait_and_close() behaviour: https://github.com/aio-libs/aiojobs/pull/546
